### ISSUE TITLE
Consider making win32 an official Flutter package

### ISSUE THREAD

**Issue by timsneath** [flutter/flutter#81222]
## Use case
Windows APIs come in a variety of flavors, most notably:

- Simple C-based APIs like [`MessageBox`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messagebox)
- COM-based APIs like [`IDesktopWallpaper`](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-idesktopwallpaper)
- WinRT APIs like [`Windows.Storage.ApplicationData`](https://docs.microsoft.com/en-us/uwp/api/windows.storage.applicationdata?view=winrt-19041)

In Flutter, we use all three types of API in various ways, not least through plugins such as [path_provider_windows](https://pub.dev/packages/path_provider_windows). That will only increase with the experimental work on the UWP runner, and the potential of moving some code from C++ or PowerShell to Dart. 

The [win32 package](https://pub.dev/packages/win32) provides a convenient low-level Dart projection of selected Windows APIs. It includes a code generator, as well as the resultant Dart library and about 1,500 unit tests. The package is [loosely endorsed by Microsoft](https://github.com/microsoft/win32metadata) as a community project. It in turn relies through a dev dependency on a separate [winmd package](https://pub.dev/packages/winmd) that can consume Windows metadata.

Both packages are maintained by @timsneath and are BSD-licensed. 

## Proposal
It does seem a little bit of an odd situation that all this code built on top of it is officially sanctioned, but not that package itself. While version pinning helps a little, it is worth considering whether we should bring win32 (at least) into the fold as an official package of the Flutter team to minimize risk. 

The specifics of this are undefined here, but might include moving the package into the flutter/plugins repo, applying a more rigorous code review, or using build infra that tests it as part of a regular Flutter Windows build.

**Comment by stuartmorgan** [comment]
> might include moving the package into the flutter/plugins repo

It would be flutter/packages, rather than flutter/plugins, since it's not a plugin.

I think there are tow conceptual arguments against that:
- It could set a precedent that the Flutter team should write/own Dart projections for all the other platforms it supports as well. In practice that seems like something we aren't likely to do in the foreseeable future, so may send a confusing message.
- It seems odd for Flutter to own a package that is just as applicable to Dart as to Flutter; a lot of the examples, in fact, don't use Flutter at all. Should we consider, e.g., https://github.com/dart-lang/win32 instead?

I don't know that these are sufficient issues to not move it, but they are worth considering.

**Comment by cbracken** [comment]
Agreed that we should move this to a more official location. Our plugins serve as an example to others, so by using it in our own, we're implicitly encouraging use across the ecosystem, so we really should have it somewhere more official. My primary concern is actually just ensuring we've placed it under a Dart/Flutter org to (a) reduce the bus factor and (b) ensure it doesn't slip into bit-rot unnoticed if Tim moves on as maintainer.

 Agreed with Stuart that this is more general than Flutter plugins though and that dart-lang seems like the right org.

**Comment by mit-mit** [comment]
Interesting question about where this logically belongs. For Dart, we've tried to focus on creating a general platform, and have actively been moving things that are framework specific (e.g. AngularDart) outside of the core Dart platform and repos. So given `Win32` really is primarily here to enable the Flutter framework, I think the most consistent would be to place it in the Flutter org.

**Comment by stuartmorgan** [comment]
> So given `Win32` really is primarily here to enable the Flutter framework

I think the question is to what extent that's true; it's equally applicable to `dart compile exe`. Anecdotally, my previous experience with similar functionality in other languages (Cocoa bindings for Python and Ruby) was all in the context of command line scripts/applications. Many `win32` APIs similarly have nothing to do with UI.

**Comment by mit-mit** [comment]
Right, is can certainly be used with the Dart framework. But that isn't sufficient to make it something the Dart team would usually create and support. If it wasn't for Flutter needing it, I suspect we'd leave `Win32` as a community package?

**Comment by stuartmorgan** [comment]
> If it wasn't for Flutter needing it, I suspect we'd leave `Win32` as a community package?

Sure, but by the same token if Tim hadn't written it then `path_provider_windows` would just have been implemented as a traditional Flutter plugin using C++, rather than via FFI, and `win32` would continue not to exist. I.e., it's not that Flutter *needs* this package, it's that this package makes FFI is significantly more useful on Windows than it would otherwise be, and so since it existed we used it. (`path_provider_macos` is a traditional plugin not because there's something fundamentally different between macOS and Windows, but because nobody happened to have written a Dart projection for Cocoa. This is basically what the first bullet point in my first comment is about)

I definitely understand the argument that the Dart team wouldn't usually create this framework, because I think it's exactly the same situation as the Flutter team.

**Comment by GroovinChip** [comment]
> It could set a precedent that the Flutter team should write/own Dart projections for all the other platforms it supports as well. In practice that seems like something we aren't likely to do in the foreseeable future, so may send a confusing message.

If I may, I'd like to suggest that this is actually a _good_ thing to do - Flutter owns material and cupertino, the go-to packages for widgets targeting mobile devices. Since Flutter supports web and desktop, I see no reason why packages targeting those platforms shouldn't be adopted by Flutter (within reason, of course, and with author consent). Tim's win32 package is a perfect example, not to mention other packages like [fluent_ui](https://pub.dev/packages/fluent_ui), [yaru](https://pub.dev/packages/yaru), and more.

**Comment by stuartmorgan** [comment]
> Flutter owns material and cupertino, the go-to packages for widgets targeting mobile devices

`win32` is not a widget package, it's an OS API language projection. They are not even remotely comparable.

**Comment by GroovinChip** [comment]
Respectfully, I feel like that's being pedantic. Yes, I used the word 'widget' and gave examples of such, but I'm trying to speak to the broader issue at hand, which is Flutter more overtly supporting desktop by way of first party libraries like win32 and others, _including_ widget libraries for desktop.

**Comment by stuartmorgan** [comment]
> Respectfully, I feel like that's being pedantic.

I do not understand how it's being pedantic to point out that owning a set of packages that do one kind of thing is not precedent for owning a completely different type of packages whose purpose is unrelated to, and entirely disjoint from, that set of packages.

> Yes, I used the word 'widget' and gave examples of such

Which is a critically important difference, not a semantic quibble about the use of a word. Widgets are a Flutter-level concept, and in fact are documented as "the central class hierarchy in the Flutter framework". It does not follow that because the Flutter teams own platform-themed implementations of the *central object type in the framework*, it should also own anything else related to any platform it runs on.

> but I'm trying to speak to the broader issue at hand, which is Flutter more overtly supporting desktop

Flutter has overtly supported mobile since its inception, and is not building and maintaining language projections for iOS and Android.
